### PR TITLE
quest(wsnm): Add WSNM Quests

### PR DESCRIPTION
**Pull request by davismj**
Adapted from https://github.com/ababyduck/darkstar-ahiru/tree/WSquests by @ababyduck.

- [x] The Walls of Your Mind (Hand-to-Hand, Asuran Fists)
- [x] The Potential Within (Great Katana, Tachi: Kasha)
- [x] Cloak and Dagger (Dagger, Evisceration)
- [x] Methods Create Madness (Polearm, Impulse Drive)
- [x] Souls in Shadow (Scythe, Spiral Hell)
- [x] Bugi Soden (Katana, Blade: Ku)
- [x] From Saplings Grow (Archery, Empyreal Arrow)
- [x] Old Wounds (Sword, Savage Blade)
- [x] The Weight of Your Limits (Great Axe, Steel Cyclone)
- [x] Axe the Competition (Axe, Decimation)
- [x] Inheritance (Great Sword, Ground Strike)
- [x] Shoot First, Ask Questions Later (Marksmanship, Detonator)
- [x] Blood and Glory (Staff, Retribution)
- [x] Orastery Woes (Club, Black Halo)

FIxes: 
- [x] Correct for merits (https://github.com/DarkstarProject/darkstar/pull/4851#pullrequestreview-120754139)
- [x] Despawn idle wsnm (https://github.com/DarkstarProject/darkstar/pull/4851#issuecomment-389557886)

### CODE REVIEW

**Comment by davismj** [comment]
This is safe to pull.

**Comment by TeoTwawki** [comment]
```
Adapted from https://github.com/ababyduck/darkstar-ahiru/tree/WSquests by @ababyduck.
```
please put this or something like it in the commit message because unlike github comments those become part of the history, and in the past we've had issues with (other) people not feeling they were properly credited and I'd like to avoid any chance of that.

File: scripts/globals/settings.lua [hunk 1]
```diff
@@ -104,14 +104,13 @@ RELIC_3RD_UPGRADE_WAIT_TIME = 3600;     -- wait time for 3rd relic upgrade (stag
...
 -- QUEST/MISSION SPECIFIC SETTINGS
-WSNM_LEVEL = 70; -- Min Level to get WSNM Quests
-WSNM_SKILL_LEVEL = 240;
 AF1_QUEST_LEVEL = 40; -- Minimum level to start AF1 quest
 AF2_QUEST_LEVEL = 50; -- Minimum level to start AF2 quest
 AF3_QUEST_LEVEL = 50; -- Minimum level to start AF3 quest
 AF1_FAME = 20; -- base fame for completing an AF1 quest
 AF2_FAME = 40; -- base fame for completing an AF2 quest
 AF3_FAME = 60; -- base fame for completing an AF3 quest
+WSNM_FAME = 60; -- base fame for completing a Weapon Skill quest
```
[the remark this answers is not in the document]
Comment (TeoTwawki) [reply scripts/globals/settings.lua]: I'm just saying skill level setting is redundant because you can easy change the skill lv in the db.

as far as the level I was saying lv for-quest sounded more useful than a setting for the fame reward.

they can all 3 get nuked.

[the remark this answers is not in the document]
Comment (TeoTwawki) [reply scripts/globals/settings.lua]: the AF fame settings make me wanna facepalm too really, but they already exist so they can stay for now, just suggesting we don't add it it

[the remark this answers is not in the document]
Comment (davismj) [reply scripts/globals/settings.lua]: Fair. Removed.

File: scripts/globals/wsquest.lua [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+        ['wsUnlockId'] = 1,
+        ['skillId'] = dsp.skill.HAND_TO_HAND,
+        ['minSkill'] = 250,
+        ['trialWeaponName'] = 'Knuckles of Trials',
+        ['trialWeaponId'] = 17507,
+        ['logId'] = BASTOK,
+        ['questId'] = THE_WALLS_OF_YOUR_MIND,
+        ['fameRegion'] = BASTOK,
+        ['wsnmName'] = 'Bodach',
+        ['wsnmId'] = 17461479,
```
Comment (zynjec) [inline scripts/globals/wsquest.lua]: I really don't like the idea of having _another_ spot we have to look for entity ids in the event they shift, however unlikely these zones are to ever see one again.

Perhaps just put them in the respective zones MobIDs.lua and in the QM script just pass it along, 

eg:
```cpp
handleQmTrigger(player, dsp.wsquest.theWallsOfYourMind, BODACH)
```

Comment (TeoTwawki) [reply scripts/globals/wsquest.lua]: I failed to notice this earlier, good thing I left it longer for more yes. That needs moved into MobIDs per zone and then getWsnmVariableName(quest) can we  `require` the correct zone's ID's dynamically by using concatenation
```lua
require("scripts/zones/" .. player:getZoneName() .. "/MobIDs";
```

Comment (TeoTwawki) [reply scripts/globals/wsquest.lua]: if it has to be there instead of the scripts in each zone I mean

File: scripts/zones/Port_Bastok/npcs/Oggbi.lua [hunk 3]
```diff
@@ -1,56 +1,57 @@
...
+    local mLvl = player:getMainLvl();
+    local mJob = player:getMainJob();
 
     if (ghostsOfThePast == QUEST_AVAILABLE and mJob == 2 and mLvl >= 40) then
         player:startEvent(231); -- Start Quest "Ghosts of the Past"
     elseif (ghostsOfThePast == QUEST_COMPLETED and player:needToZone() == false and theFirstMeeting == QUEST_AVAILABLE and mJob == 2 and mLvl >= 50) then
         player:startEvent(233); -- Start Quest "The First Meeting"
     elseif (player:hasKeyItem(dsp.ki.LETTER_FROM_DALZAKK) and player:hasKeyItem(dsp.ki.SANDORIAN_MARTIAL_ARTS_SCROLL)) then
         player:startEvent(234); -- Finish Quest "The First Meeting"
+    elseif (wsQuestEvent ~= nil) then
```
Comment (zynjec) [inline scripts/zones/Port_Bastok/npcs/Oggbi.lua]: Kind of pedantic, but based on wiki guide for this quest suggest it WSNM quests are at the top of the food chain, which for us is just making sure it's the first condition in the if statement for now.

I kind of assume if this is true, it'd be the case for all other WSNM quests.

http://ffxiclopedia.wikia.com/wiki/Ghosts_of_the_Past
```Note: If you have hand-to-hand skill above 250 and he offers you the Knuckles of Trials for the quest The Walls of Your Mind you will not be able to start this quest until you've completed the trial.```

File: scripts/globals/wsquest.lua [hunk 4]
```diff
@@ -0,0 +1,634 @@
...
+    --------------------------------------
+    ['theWallsOfYourMind'] =
+    {
+        ['questName'] = 'The Walls Of Your Mind',
+        ['wsName'] = 'Asuran Fists',
+        ['wsId'] = 9,
+        ['wsUnlockId'] = 1,
+        ['skillId'] = dsp.skill.HAND_TO_HAND,
+        ['minSkill'] = 250,
+        ['trialWeaponName'] = 'Knuckles of Trials',
```
Comment (zynjec) [inline scripts/globals/wsquest.lua]: Same thing, just describing something already in comment block.

File: scripts/globals/wsquest.lua [hunk 5]
```diff
@@ -0,0 +1,634 @@
...
+dsp.wsquest =
+{
+    --------------------------------------
+    -- Quest: The Walls of Your Mind
+    --   NPC: Oggbi (Port Bastok)
+    --    WS: Asuran Fists (Hand-to-Hand)
+    --------------------------------------
+    ['theWallsOfYourMind'] =
+    {
+        ['questName'] = 'The Walls Of Your Mind',
```
Comment (zynjec) [inline scripts/globals/wsquest.lua]: Is there even any reason to have this? Not used anywhere, and the comments above each block explain which combat skill and weapon it's for.

File: scripts/globals/wsquest.lua [hunk 6]
```diff
@@ -0,0 +1,634 @@
...
+        ['wsName'] = 'Asuran Fists',
```
Comment (zynjec) [inline scripts/globals/wsquest.lua]: Same thing, just describing something already in comment block.

File: scripts/globals/wsquest.lua [hunk 7]
```diff
@@ -0,0 +1,634 @@
...
+            SpawnMob(quest.wsnmId):updateClaim(player);
+        end;
+    else
+        player:messageSpecial(NOTHING_OUT_OF_ORDINARY);
+    end;
+end;
+
+function handleWsnmDeath(quest, player)
+    if (getWsQuestState(quest, player) == WSQUEST_CONT2) then
+        player:setVar(getWsnmVariableName(quest), 1);
```
Comment (zynjec) [inline scripts/globals/wsquest.lua]: If we're keeping this system as is, I think this would be a perfect use case for local variable, instead of a character variable.

It makes a very temporary (destroyed on zone), and makes it more retail accurate, because if you zone without picking up the key item, you have to kill NM again.

File: scripts/globals/wsquest.lua [hunk 8]
```diff
@@ -0,0 +1,634 @@
...
+    end;
+
+    return nil;
+end;
+
+local function getWsnmVariableName(quest)
+    return string.format('Killed_%s', string.gsub(quest.wsnmName, ' ', '_'));
+end;
+
+function getWsQuestTradeEvent(quest, player, trade)
```
Comment (takhlaq) [inline scripts/globals/wsquest.lua]:
```lua
dsp.wsquest.getTradeEvent = function(quest, player, trade)
```

File: scripts/globals/wsquest.lua [hunk 9]
```diff
@@ -0,0 +1,634 @@
...
+            return quest.eventIds.tradedUnfinishedWeapon;
+        else
+            return quest.eventIds.tradedFinishedWeapon;
+        end;
+    end;
+
+    return nil;
+end;
+
+function getWsQuestTriggerEvent(quest, player)
```
Comment (takhlaq) [inline scripts/globals/wsquest.lua]:
```lua
dsp.wsquest.getTriggerEvent = function(quest, player)
```

File: scripts/globals/wsquest.lua [hunk 10]
```diff
@@ -0,0 +1,634 @@
...
+    elseif (state == WSQUEST_CONT2) then
+        return quest.eventIds.cont2;
+    elseif (state == WSQUEST_FINISH) then
+        return quest.eventIds.finish; -- WS Quest completed (WSNM killed)
+    end;
+
+    return nil;
+end;
+
+function handleQmTrigger(quest, player)
```
Comment (takhlaq) [inline scripts/globals/wsquest.lua]:
```lua
dsp.wsquest.handleQmTrigger = function(quest, player)
```

File: scripts/globals/wsquest.lua [hunk 11]
```diff
@@ -0,0 +1,634 @@
...
+        elseif (GetMobAction(quest.wsnmId) == 0) then
+            player:messageSpecial(SENSE_OMINOUS_PRESENCE);
+            SpawnMob(quest.wsnmId):updateClaim(player);
+        end;
+    else
+        player:messageSpecial(NOTHING_OUT_OF_ORDINARY);
+    end;
+end;
+
+function handleWsnmDeath(quest, player)
```
Comment (takhlaq) [inline scripts/globals/wsquest.lua]:
```lua
dsp.wsquest.handleWsnmDeath(quest, player)
```

File: scripts/globals/wsquest.lua [hunk 12]
```diff
@@ -0,0 +1,634 @@
...
+    if (getWsQuestState(quest, player) == WSQUEST_CONT2) then
+        player:setVar(getWsnmVariableName(quest), 1);
+    end;
+end;
+
+function handleWsQuestEventFinish(quest, player, csid, option, learnedId)
```
Comment (takhlaq) [inline scripts/globals/wsquest.lua]:
```lua
dsp.wsquest.handleEventFinish(quest, player, csid, option, learnedId)
```

File: scripts/globals/wsquest.lua (hunk 4 above)
Comment (davismj) [reply scripts/globals/wsquest.lua]: I debated removing it as well.

File: scripts/globals/wsquest.lua (hunk 7 above)
Comment (davismj) [reply scripts/globals/wsquest.lua]: local to which block?

Comment (zynjec) [reply scripts/globals/wsquest.lua]: The lua entity binding setLocalVar: https://github.com/DarkstarProject/darkstar/blob/master/src/map/lua/lua_baseentity.cpp#L541

The variable is only temporarily stored with the entity, and not stored in the database. It would be destroyed when the player zones.

File: scripts/globals/wsquest.lua [hunk 13]
```diff
@@ -20,6 +20,7 @@ dsp.wsquest =
         ['logId'] = BASTOK,
         ['questId'] = THE_WALLS_OF_YOUR_MIND,
         ['fameRegion'] = BASTOK,
+        ['wsnmName'] = 'Bodach',
```
Comment (zynjec) [inline scripts/globals/wsquest.lua]: Why are these needed at all?

You can only have one trial active at a time, so it's not like you'll have a collision. Just use a generic `killed_wsnm` variable name or something and do away with getvariablename/wsnmName all together imo.

File: sql/mob_groups.sql [hunk 14]
```diff
@@ -8823,7 +8823,7 @@ INSERT INTO `mob_groups` VALUES (8771,233,167,7920,0,3445,0,0,56,62,0);
 INSERT INTO `mob_groups` VALUES (8772,461,167,1056,0,334,0,0,65,65,0);
 INSERT INTO `mob_groups` VALUES (8773,482,167,7920,128,0,7500,0,80,80,0);
 INSERT INTO `mob_groups` VALUES (8774,908,167,1056,0,636,0,0,52,54,0);
-INSERT INTO `mob_groups` VALUES (8775,1118,167,7920,128,798,0,0,72,73,0);
+INSERT INTO `mob_groups` VALUES (8775,1118,167,0,128,798,0,0,72,73,0);
```
Comment (zynjec) [inline sql/mob_groups.sql]: Looks like you hit `Drexerion the Condemned` instead of `Bodach`

Comment (davismj) [reply sql/mob_groups.sql]: lol wtf

**Comment by davismj** [comment]
@zynjec So the data from here has levels other than 80. Would you be able to confirm that one of the non-80's listed is in fact 80? 

https://github.com/ababyduck/darkstar-ahiru/projects/2

**Comment by zynjec** [comment]
I can see about doing a non-80 one on that list at some point in the next 2 weeks.

I also had another thought, there should probably also be a `IDLE_DESPAWN` modmob on the WSNM as well, in the event someone actually dies to it.

File: scripts/globals/wsquest.lua [hunk 15]
```diff
@@ -0,0 +1,562 @@
...
+        if (player:hasKeyItem(dsp.ki.ANNALS_OF_TRUTH)) then
+            return WSQUEST_FINISH
+        elseif (player:hasKeyItem(dsp.ki.MAP_TO_THE_ANNALS_OF_TRUTH)) then
+            return WSQUEST_CONT2
+        else
+            return WSQUEST_CONT1
+        end
+    else
+        local canEquip = player:canEquipItem(quest.trialWeaponId,true)
+        local sufficientSkill = player:getSkillLevel(quest.skillId) >= quest.minSkill
```
Comment (Hozu) [inline scripts/globals/wsquest.lua]: Doesn't getSkillLevel return the current skill value, ie. including merits and other modifiers? Actually, does retail check for base skill level or current for WS quests? If it's the former, IIRC there's a function that returns the base skill level, but I can't remember what it was.

Comment (zynjec) [reply scripts/globals/wsquest.lua]: You're right, getSkillLevel, I forgot it returns the total skill the character has (merits, gear, etc).

I just tested this on retail with Great Axe, native skill is 228 but with merits I have 240 effective skill, and Iron Eater will not give me the quest, so we'll have to change this.

File: scripts/globals/wsquest.lua [hunk 16]
```diff
@@ -0,0 +1,562 @@
...
+
+dsp = dsp or {}
+dsp.wsquest =
+{
+    --------------------------------------
+    -- Quest: The Walls of Your Mind
+    --   NPC: Oggbi (Port Bastok)
+    --    WS: Asuran Fists (Hand-to-Hand)
+    --------------------------------------
+    ['theWallsOfYourMind'] =
```
Comment (zynjec) [inline scripts/globals/wsquest.lua]: Just thinking out loud here. I am wondering if this would be a bit more intuitive to rename these to either the combat skill itself, or the weapon skill it unlocks, 

`dsp.wsquest.theWallsOfYourMind` really isn't very obvious what it's for without some digging, where something like `dsp.wsquest.hand_to_hand` or `dsp.wsquest.asuran_fists` is pretty obvious.

File: scripts/commands/setplayerskill.lua [hunk 17]
```diff
@@ -0,0 +1,48 @@
+---------------------------------------------------------------------------------------------------
+-- func: setplayerlevel
```
Comment (nfaletra) [inline scripts/commands/setplayerskill.lua]: This function name does not accurately describe what the command is doing.

File: scripts/globals/wsquest.lua [hunk 18]
```diff
@@ -0,0 +1,562 @@
...
+dsp.wsquest =
+{
+    --------------------------------------
+    -- Quest: The Walls of Your Mind
+    --   NPC: Oggbi (Port Bastok)
+    --    WS: Asuran Fists (Hand-to-Hand)
+    --------------------------------------
+    ['asuran_fists'] =
+    {
+        ['wsId'] = 9,
```
Comment (nfaletra) [inline scripts/globals/wsquest.lua]: weaponskillids.lua have these stored off so we can get rid of hardcoded values.

File: scripts/zones/Kazham/TextIDs.lua [hunk 19]
```diff
@@ -29,3 +29,6 @@ GHEMISENTERILO_SHOP_DIALOG = 10042; -- Can you really get everything that you wa
 
 -- Porter Moogle
         RETRIEVE_DIALOG_ID = 10994; -- You retrieve$ from the porter moogle's care.
+
+-- WS Quest
+      EVISCERATION_LEARNED = 10644; -- You have learned the weapon skill "Evisceration"!
```
Comment (nfaletra) [inline scripts/zones/Kazham/TextIDs.lua]: This TextID is off as of the May 2018 Update. Should be 10645.

File: scripts/zones/Norg/TextIDs.lua [hunk 20]
```diff
@@ -37,3 +37,7 @@ YOU_CAN_NOW_BECOME_A_SAMURAI = 10196; -- You can now become a samurai.
 
 -- Porter Moogle
           RETRIEVE_DIALOG_ID = 11274; -- You retrieve$ from the porter moogle's care.
+
+-- WS Quest
+         TACHI_KASHA_LEARNED = 10796; -- You have learned the weapon skill “Tachi: Kasha”!
```
Comment (nfaletra) [inline scripts/zones/Norg/TextIDs.lua]: 10797

File: scripts/zones/Norg/TextIDs.lua [hunk 21]
```diff
@@ -37,3 +37,7 @@ YOU_CAN_NOW_BECOME_A_SAMURAI = 10196; -- You can now become a samurai.
...
+            BLADE_KU_LEARNED = 10821; -- You have learned the weapon skill “Blade: Ku”!
```
Comment (nfaletra) [inline scripts/zones/Norg/TextIDs.lua]: 10822

File: scripts/zones/Port_Bastok/TextIDs.lua [hunk 22]
```diff
@@ -18,7 +18,7 @@
 
 -- Mission Dialogs
      YOU_ACCEPT_THE_MISSION = 7339; -- You have accepted the mission.
-    ORIGINAL_MISSION_OFFSET = 7344; -- You can consult the ission section of the main menu to review your objectives. Speed and efficiency are your priorities. Dismissed. 
+    ORIGINAL_MISSION_OFFSET = 7344; -- You can consult the ission section of the main menu to review your objectives. Speed and efficiency are your priorities. Dismissed.
```
Comment (nfaletra) [inline scripts/zones/Port_Bastok/TextIDs.lua]: 7345

File: scripts/zones/Port_Bastok/TextIDs.lua [hunk 23]
```diff
@@ -30,7 +30,7 @@
            NOKKHI_BAD_COUNT = 8787; -- What kinda smart-alecky baloney is this!? I told you to bring me the same kinda ammunition in complete sets. And don't forget the flowers, neither.
           NOKKHI_GOOD_TRADE = 8789; -- And here you go! Come back soon, and bring your friends!
             NOKKHI_BAD_ITEM = 8790; -- I'm real sorry, but there's nothing I can do with those.
-            CLOUD_BAD_COUNT = 9214; -- Well, don't just stand there like an idiot! I can't do any bundlin' until you fork over a set of 99 tools and 
+            CLOUD_BAD_COUNT = 9214; -- Well, don't just stand there like an idiot! I can't do any bundlin' until you fork over a set of 99 tools and
```
Comment (nfaletra) [inline scripts/zones/Port_Bastok/TextIDs.lua]: 9215

File: scripts/zones/Port_Bastok/TextIDs.lua [hunk 24]
```diff
@@ -74,3 +74,6 @@ DHENTEVRYUKOH_CLOSED_DIALOG = 8363; -- I'm trrrying to start a business selling
 
 -- conquest Base
               CONQUEST_BASE = 6524; -- Tallying conquest results...
+
+-- WS Quest
+       ASURAN_FISTS_LEARNED = 8806; -- You have learned the weapon skill “Asuran Fists”!
```
Comment (nfaletra) [inline scripts/zones/Port_Bastok/TextIDs.lua]: 8807

File: scripts/zones/Port_Bastok/TextIDs.lua (hunk 23 above)
Comment (nfaletra) [reply scripts/zones/Port_Bastok/TextIDs.lua]: Rebasing should fix this one.

File: scripts/zones/Port_Bastok/TextIDs.lua (hunk 22 above)
Comment (nfaletra) [reply scripts/zones/Port_Bastok/TextIDs.lua]: Rebasing should fix this one.

File: scripts/zones/Bostaunieux_Oubliette/TextIDs.lua [hunk 25]
```diff
@@ -14,3 +14,8 @@ ITEM_CANNOT_BE_OBTAINED = 6539; -- You cannot obtain the item <item>. Come back
 
 -- conquest Base
           CONQUEST_BASE =    0;
+
+-- WS Quest
+    SPIRAL_HELL_LEARNED = 7415; -- You have learned the weapon skill “Spiral Hell”!
+ SENSE_OMINOUS_PRESENCE = 7416; -- You sense an ominous presence...
+NOTHING_OUT_OF_ORDINARY = 7417; -- There is nothing out of the ordinary here
```
[the remark this answers is not in the document]
Comment (nfaletra) [reply scripts/zones/Bostaunieux_Oubliette/TextIDs.lua]: Increment by one. 7416, 7417, 7418

File: scripts/zones/Gustav_Tunnel/TextIDs.lua [hunk 26]
```diff
@@ -9,5 +9,7 @@ ITEM_CANNOT_BE_OBTAINED = 6380; -- You cannot obtain the item <item> come back a
 
 -- conquest Base
           CONQUEST_BASE = 7047; -- Tallying conquest results...
-
 NOTHING_OUT_OF_ORDINARY = 7307; -- There is nothing out of the ordinary here
+
+-- WS Quest
+ SENSE_OMINOUS_PRESENCE = 7306; -- You sense an ominous presence...
```
[the remark this answers is not in the document]
Comment (nfaletra) [reply scripts/zones/Gustav_Tunnel/TextIDs.lua]: 7307

File: scripts/zones/Kuftal_Tunnel/TextIDs.lua [hunk 27]
```diff
@@ -1,29 +1,32 @@
...
+                   FELL = 7331 -- The piece of wood fell off the cliff!
+                   EVIL = 7332 -- You sense an evil presence...
+              FISHBONES = 7346 -- Fish bones lie scattered about the area...
 
 -- conquest Base
-          CONQUEST_BASE = 7047; -- Tallying conquest results...
+          CONQUEST_BASE = 7047 -- Tallying conquest results...
+
+-- WS Quest
+ SENSE_OMINOUS_PRESENCE = 7348 -- You sense an ominous presence...
```
[the remark this answers is not in the document]
Comment (nfaletra) [reply scripts/zones/Kuftal_Tunnel/TextIDs.lua]: 7349. . .I'm sensing a pattern.

**Comment by TeoTwawki** [comment]
Lets put the GM commands to a separate pull request, per discussion in discord direct message

File: scripts/globals/wsquest.lua (hunk 15 above)
Comment (Hozu) [reply scripts/globals/wsquest.lua]: Took a look around, I believe the function you'll need is getCharSkillLevel()

File: scripts/zones/The_Boyahda_Tree/npcs/qm1.lua [hunk 28]
```diff
@@ -0,0 +1,15 @@
+-----------------------------------
+-- Area: The Boyahda Tree (153)
+--  NPC: qm1 (???)
+-- Quests: Shoot First, Ask Questions Later (Detonator WSNM "Beet Leafhopper")
+-- !pos -11 -19 -177 153
+-----------------------------------
+package.loaded["scripts/zones/The_Boyahda_Tree/TextIDs"] = nil;
```
Comment (takhlaq) [inline scripts/zones/The_Boyahda_Tree/npcs/qm1.lua]: 👁
